token/link/link.go

```diff
@@ -0,0 +1,233 @@
+	identityUUID, err := uuid.FromString(identityID)
+	if err != nil {
+		return "", err
+	}
+
+	forResource := referrerURL.Query().Get(forParam)
+
+	oauthProvider, err := service.providerFactory.NewOauthProvider(ctx, req, service.config, forResource)
+	if err != nil {
+		return "", err
+	}
+
+	providerToken, err := oauthProvider.Exchange(ctx, code)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"state": state,
+			"code":  code,
+			"err":   err,
+		}, "exchange operation failed")
+		return "", err
+	}
+	if providerToken.AccessToken == "" {
+		log.Error(ctx, map[string]interface{}{
+			"state":       state,
+			"code":        code,
+			"provider_id": oauthProvider.ID(),
+		}, "access token return by provider is empty")
+		return "", errors.New("access token return by provider is empty")
+	}
+
+	err = application.Transactional(service.db, func(appl application.Application) error {
+		tokens, err := appl.ExternalTokens().LoadByProviderIDAndIdentityID(ctx, oauthProvider.ID(), identityUUID)
+		if err != nil {
+			return err
+		}
+		if len(tokens) > 0 {
+			// It was re-linking. Overwrite the existing link.
+			externalToken := tokens[0]
+			externalToken.Token = providerToken.AccessToken
+			err = appl.ExternalTokens().Save(ctx, &externalToken)
+			if err == nil {
+				log.Info(ctx, map[string]interface{}{
+					"provider_id":       oauthProvider.ID(),
+					"identity_id":       identityID,
+					"external_token_id": externalToken.ID,
+				}, "An existing token found. Account re-linked & new token saved.")
+			}
+			return err
+		}
+		externalToken := provider.ExternalToken{
+			Token:      providerToken.AccessToken,
+			IdentityID: identityUUID,
+			Scope:      oauthProvider.Scopes(),
+			ProviderID: oauthProvider.ID(),
+		}
+		err = appl.ExternalTokens().Create(ctx, &externalToken)
+		if err == nil {
+			log.Info(ctx, map[string]interface{}{
+				"provider_id":       oauthProvider.ID(),
+				"identity_id":       identityID,
+				"external_token_id": externalToken.ID,
+			}, "No old token found. Account linked & new token saved.")
+		}
+		return err
+	})
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"state":       state,
+			"code":        code,
+			"provider_id": oauthProvider.ID(),
+			"identity_id": identityID,
+		}, "failed to save token")
+		return "", err
+	}
+
+	return knownReferrer, nil
+}
+
+// NewOauthProvider creates a new oauth provider for the given resource URL
+func (service *LinkService) NewOauthProvider(ctx context.Context, req *goa.RequestData, config LinkConfig, forResource string) (ProviderConfig, error) {
+	authURL := rest.AbsoluteURL(req, "")
+
+	resourceURL, err := url.Parse(forResource)
+	if err != nil {
+		return nil, err
+	}
+	if resourceURL.Host == "github.com" {
+		return NewGitHubConfig(config.GetGitHubClientID(), config.GetGitHubClientSecret(), config.GetGitHubClientDefaultScopes(), authURL), nil
+	} else if strings.HasPrefix(forResource, config.GetOpenShiftClientApiUrl()) {
```

Comment:
This implies we are supporting : 
```
`for=https://<IP_Address>` 
```

Would be handy for local setups using minishift!

Reply:
Yeah... this was intended :)

Reply:
This worked!